ntoskrnl/mm/ARM3/miarm.h

```diff
@@ -2387,13 +2409,107 @@ MiRosUnmapViewInSystemSpace(
     IN PVOID MappedBase
 );
 
+FORCEINLINE
 VOID
-NTAPI
 MiMakePdeExistAndMakeValid(
-    IN PMMPDE PointerPde,
-    IN PEPROCESS TargetProcess,
-    IN KIRQL OldIrql
-);
+    _In_ PMMPDE PointerPde,
+    _In_ PEPROCESS TargetProcess)
+{
+    NTSTATUS Status = MiUserAccessFaultUnsafe(0, TargetProcess, MiPdeToPte(PointerPde), KernelMode, NULL);
+
+    /* Caller must check that this is a valid call. */
+    if (!NT_SUCCESS(Status))
+    {
+        KeBugCheckEx(KERNEL_DATA_INPAGE_ERROR,
+                     1,
+                     Status,
+                     (ULONG_PTR)TargetProcess,
+                     (ULONG_PTR)MiPdeToPte(PointerPde));
+    }
+}
+
+
+/*
+ * We put barriers here because we want the compiler
+ * to really emit checks where we put them. (e.g after making the PPE valid).
+ * We also put barriers before returning, because we want the compiler to
+ * put memory access instructions (e.g PTEs in the verified page directory) after those checks are made.
+ */
+
+FORCEINLINE
+BOOLEAN
+MiIsPteValid(PMMPTE PointerPte)
+{
+    BOOLEAN ret;
+    KeMemoryBarrierWithoutFence();
+    ret = PointerPte->u.Hard.Valid;
+    KeMemoryBarrierWithoutFence();
```

Comment:
Looks like we're trying to emulate `atomic_load()` from C11 here

Reply:
Doesn't `atomic_load` involve some specific CPU instruction ? Here the Memory Barrier is for compiler only.

Reply:
On x86, it doesn't do anything special except being a compiler barrier
https://godbolt.org/z/h6Yo7Y4z5